src/renderers/common/StorageBufferAttribute.js

```diff
@@ -2,14 +2,18 @@ import { BufferAttribute } from '../../core/BufferAttribute.js';
 
 class StorageBufferAttribute extends BufferAttribute {
 
-	constructor( array, itemSize, typeClass = Float32Array ) {
+	constructor( array, itemSize, indirect = false, typeClass = Float32Array ) {
```

Comment:
Hmm.. I  thought extends `IndirectStorageBufferAttribute` from `StorageBufferAttribute`. https://github.com/mrdoob/three.js/pull/29594#discussion_r1792587593

Changing the constructor signature can be confusing, and having a class to handle indirect buffering can even help with reading the code among some other minor optimizations.

Reply:
Agreed. should remove the changes in StorageBufferAttribute and simply create a new file `IndirectStorageBufferAttribute`:
```js
import { StorageBufferAttribute } from '../../core/StorageBufferAttribute.js';

class IndirectStorageBufferAttribute extends StorageBufferAttribute {

	constructor( array, itemSize, typeClass = Float32Array ) {

		super( array, itemSize, typeClass );

		this.isIndirectStorageBufferAttribute = true;

	}

}

export default IndirectStorageBufferAttribute;
```

Reply:
ah ok, on this occasion I recommend as a distinguishing criterion to use

"this.isIndirect = false;" in StorageBufferAttribute.js
"this.isIndirect = true;" in IndirectStorageBufferAttribute.js

The reason I did this is because I didn't like the duplication of StorageBuffer.js and NodeStorageBuffer.js. I first doubled this because StorageBuffer.js had no distinguishing criterion which is important for later queries. 
This means that with isStorageBufferAttribute in both classes ( StorageBufferAttribute / IndirectStorageBufferAttribute ) we have the opportunity to query anywhere in the code whether it is a storageBufferAttribute and with isIndirect whether it is an indirect storageBuffer. This makes things a lot easier because unnecessary duplication is not necessary in all places where it is only necessary to check whether there is a storageBuffer and / or indirect.

@sunag I don't have a simple example at the moment because I'm testing in my virtual geometry environment. But I have something in mind. The simplest possible example would be an instancedGeometry in which individual instances are randomly made invisible or visible over time with a sin function and a threshold value of 0.5 in the compute shader.
I looked at your reference to the draw function. I'll have to take a closer look at this because so far I've focused on the part with the visibility test. I think it makes sense if the geometries as renderObjects carry the drawBuffer as an attribute if you want to use drawIndirect. This makes it easy to distinguish between a mixture of many geometries that use draw and others that use drawIndirect.

@RenaudRohlinger Thanks for your link, it helped me understand renderBundles. I think if we get both combined in the end we can pop the corks

Suggestions for improvement are welcome

Reply:
> ah ok, on this occasion I recommend as a distinguishing criterion to use
"this.isIndirect = false;" in StorageBufferAttribute.js
"this.isIndirect = true;" in IndirectStorageBufferAttribute.js

The @RenaudRohlinger example already does this just once it has the `isIndirectStorageBufferAttribute=true` variable in `IndirectStorageBufferAttribute`, this will always be missing (undefined) in `StorageBufferAttribute`.

Reply:
@Spiri0 I made some revisions and I'm going to merge it up, I'll see if I implement these encoders separately, I want to see if I make a simple example too so we can have them in the official examples.

Reply:
I like it, it became pretty beauty in the end. As a user, I was in awe of tackle so many critical parts of the node system. The topic helped me to understand it much better. It is a very impressive construction.

I will now take a closer look at the custom structs that I have in mind. The big advantage of this will be instead of e.g. 30 individual bindings only need 4 or 5. This reduces overhead, pools cache and makes the shader more compact.

As for the encoder topic. I had looked at the geometries class in the node system because I had in mind to pass the drawBuffers as an attribute with the geometries as a renderObject, but it seams you have a better solution in mind 👍